File: app/src/main/java/com/nirbhor/app/ui/screens/record/RecordScreen.kt
package com.nirbhor.app.ui.screens

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.nirbhor.app.data.NirbhorRepository
import com.nirbhor.app.navigation.LocalAppContainer
import com.nirbhor.app.navigation.NavActions
import com.nirbhor.app.ui.components.NbCard
import com.nirbhor.app.ui.components.ProgressRing
import com.nirbhor.app.ui.components.SecondaryButton
import com.nirbhor.app.ui.components.SegmentedControl
import com.nirbhor.app.ui.i18n.LocalIsBangla
import com.nirbhor.app.ui.i18n.num
import com.nirbhor.app.ui.i18n.percent
import com.nirbhor.app.ui.i18n.tr
import com.nirbhor.app.ui.marks.MedicineMark
import com.nirbhor.app.ui.theme.Dimens
import com.nirbhor.app.ui.theme.NirbhorTheme

/** Adherence record (2t/2f): donut summary, day grid, per-medicine bars, PDF export. */
@Composable
fun RecordScreen(actions: NavActions) {
    val colors = NirbhorTheme.colors
    val repo = LocalAppContainer.current.repository
    var weekly by remember { mutableStateOf(true) }
    val days = if (weekly) 7 else 30

    var window by remember { mutableStateOf<NirbhorRepository.AdherenceWindow?>(null) }
    var cells by remember { mutableStateOf<List<NirbhorRepository.DayCell>>(emptyList()) }
    var perMed by remember { mutableStateOf<List<NirbhorRepository.MedAdherence>>(emptyList()) }
    LaunchedEffect(days) {
        window = repo.adherenceOver(days)
        cells = repo.dayCells(days)
        perMed = repo.perMedicineAdherence(days)
    }

    Column(Modifier.fillMaxSize().background(colors.paper)) {
        Row(
            Modifier.fillMaxWidth().windowInsetsPadding(WindowInsets.statusBars)
                .padding(horizontal = Dimens.screenPadding, vertical = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(tr("রেকর্ড", "Record"), style = NirbhorTheme.type.titleHero, color = colors.ink, modifier = Modifier.weight(1f))
            Box(Modifier.width(160.dp)) {
                SegmentedControl(
                    options = listOf(tr("সপ্তাহ", "Week"), tr("মাস", "Month")),
                    selectedIndex = if (weekly) 0 else 1,
                    onSelect = { weekly = it == 0 },
                )
            }
        }
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState())
                .padding(horizontal = Dimens.screenPadding).padding(bottom = 24.dp),
            verticalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            SummaryCard(window)
            DayGrid(cells)
            perMed.filter { it.total > 0 }.forEach { MedBar(it) }
            SecondaryButton(
                tr("ডাক্তারের জন্য পিডিএফ বানান", "Make a PDF for the doctor"),
                actions::openDoctorReport, height = 50.dp, modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

@Composable
private fun SummaryCard(w: NirbhorRepository.AdherenceWindow?) {
    val colors = NirbhorTheme.colors
    NbCard(padding = 13.dp) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            ProgressRing(
                fraction = (w?.percent ?: 0) / 100f, diameter = 74.dp, strokeWidth = 8.dp,
                center = { Text(percent(w?.percent ?: 0), fontSize = 20.sp, fontWeight = FontWeight(700), color = colors.calmD) },
            )
            Spacer(Modifier.width(16.dp))
            Column(Modifier.weight(1f)) {
                val pct = w?.percent ?: 0
                Text(
                    if (pct >= 90) tr("বেশ ভালো চলছে", "Going well")
                    else if (pct >= 70) tr("মোটামুটি চলছে", "Fairly steady")
                    else tr("আরও যত্ন দরকার", "Needs more care"),
                    style = NirbhorTheme.type.cardTitlePrimary, color = colors.ink,
                )
                Text(
                    if ((w?.missed ?: 0) == 0) tr("এই সময়ে কোনো ডোজ বাদ পড়েনি।", "No doses were missed in this period.")
                    else tr("এই সময়ে ${num(w?.missed ?: 0)}টি ডোজ বাদ পড়েছে।", "${w?.missed ?: 0} doses were missed in this period."),
                    style = NirbhorTheme.type.meta, color = colors.ink2,
                )
            }
        }
    }
}

private val bnWeekHeaders = listOf("সোম", "মঙ্গল", "বুধ", "বৃহ", "শুক্র", "শনি", "রবি")
private val enWeekHeaders = listOf("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")

@Composable
private fun DayGrid(cells: List<NirbhorRepository.DayCell>) {
    val colors = NirbhorTheme.colors
    val bangla = LocalIsBangla.current
    val headers = if (bangla) bnWeekHeaders else enWeekHeaders
    val today = java.time.LocalDate.now()

    NbCard(padding = 13.dp) {
        Row(Modifier.fillMaxWidth()) {
            headers.forEach { h ->
                Box(Modifier.weight(1f), contentAlignment = Alignment.Center) {
                    Text(h, fontSize = 12.sp, color = colors.ink3)
                }
            }
        }
        Spacer(Modifier.height(6.dp))
        // Chunk cells into weeks of 7. Pad the first week so weekday columns align (Mon=0).
        val firstDow = cells.firstOrNull()?.date?.dayOfWeek?.value?.minus(1) ?: 0
        val padded = List(firstDow) { null } + cells
        padded.chunked(7).forEach { week ->
            Row(Modifier.fillMaxWidth().padding(vertical = 3.dp), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                for (i in 0 until 7) {
                    val cell = week.getOrNull(i)
                    Box(Modifier.weight(1f).aspectRatio(1f)) {
                        if (cell != null) DayCellView(cell, isToday = cell.date == today)
                    }
                }
            }
        }
        Spacer(Modifier.height(10.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            LegendSwatch(colors.calm, tr("সব নেওয়া", "All taken"))
            LegendSwatch(colors.warm, tr("বাদ", "Missed"))
            LegendSwatch(colors.sage, tr("বাকি", "Future"))
        }
    }
}

@Composable
private fun DayCellView(cell: NirbhorRepository.DayCell, isToday: Boolean) {
    val colors = NirbhorTheme.colors
    val shape = RoundedCornerShape(9.dp)
    Box(
        Modifier.fillMaxSize().clip(shape)
            .background(
                when (cell.state) {
                    NirbhorRepository.DayState.FULL -> colors.calm
                    NirbhorRepository.DayState.MISSED -> colors.warm
                    else -> colors.sage
                }
            ),
    ) {
        if (cell.state == NirbhorRepository.DayState.PARTIAL) {
            // 135° diagonal split: calm top-left triangle over the sage base.
            Canvas(Modifier.fillMaxSize()) {
                val p = Path().apply {
                    moveTo(0f, 0f); lineTo(size.width, 0f); lineTo(0f, size.height); close()
                }
                drawPath(p, colors.calm)
            }
        }
        if (isToday) {
            Canvas(Modifier.fillMaxSize()) {
                drawRoundRect(
                    color = colors.ink, style = Stroke(width = 2.5.dp.toPx()),
                    topLeft = Offset(-2.dp.toPx(), -2.dp.toPx()),
                    size = androidx.compose.ui.geometry.Size(size.width + 4.dp.toPx(), size.height + 4.dp.toPx()),
                    cornerRadius = androidx.compose.ui.geometry.CornerRadius(9.dp.toPx(), 9.dp.toPx()),
                )
            }
        }
    }
}

@Composable
private fun LegendSwatch(color: Color, label: String) {
    val colors = NirbhorTheme.colors
    Row(verticalAlignment = Alignment.CenterVertically) {
        Box(Modifier.size(12.dp).clip(RoundedCornerShape(4.dp)).background(color))
        Spacer(Modifier.width(5.dp))
        Text(label, style = NirbhorTheme.type.meta, color = colors.ink3)
    }
}

@Composable
private fun MedBar(m: NirbhorRepository.MedAdherence) {
    val colors = NirbhorTheme.colors
    val good = m.percent >= 80
    Row(verticalAlignment = Alignment.CenterVertically) {
        MedicineMark(shape = m.medicine.mark, color = Color(m.medicine.markColor), size = 20.dp)
        Spacer(Modifier.width(10.dp))
        Text(m.medicine.displayName, style = NirbhorTheme.type.meta, color = colors.ink2, modifier = Modifier.width(96.dp))
        Spacer(Modifier.width(8.dp))
        Box(Modifier.weight(1f).height(9.dp).clip(RoundedCornerShape(5.dp)).background(colors.sage)) {
            Box(
                Modifier.fillMaxWidth(m.percent / 100f).height(9.dp).clip(RoundedCornerShape(5.dp))
                    .background(if (good) colors.calm else colors.warm),
            )
        }
        Spacer(Modifier.width(8.dp))
        Text(percent(m.percent), style = NirbhorTheme.type.meta, color = colors.ink2)
    }
}

File: lib/ui/screens/record_screen.dart
import 'package:flutter/material.dart';

import '../../data/app_scope.dart';
import '../../data/repository.dart';
import '../../domain/dose_scheduler.dart';
import '../../navigation/nav_actions.dart';
import '../../theme/theme.dart';
import '../components/buttons.dart';
import '../components/controls.dart';
import '../components/progress_ring.dart';
import '../components/surfaces.dart';
import '../marks/medicine_mark.dart';

/// Adherence record (2t/2f) — the donut summary, the day grid, and the per-medicine bars.
class RecordScreen extends StatefulWidget {
  const RecordScreen({super.key});

  @override
  State<RecordScreen> createState() => _RecordScreenState();
}

class _RecordScreenState extends State<RecordScreen> {
  bool _weekly = true;

  int get _days => _weekly ? 7 : 30;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return ColoredBox(
      color: colors.paper,
      child: Column(
        children: [
          SafeArea(
            bottom: false,
            child: Padding(
              padding: const EdgeInsets.symmetric(
                horizontal: Dimens.screenPadding,
                vertical: 14,
              ),
              child: Row(
                children: [
                  Expanded(
                    child: Text(
                      context.tr('রেকর্ড', 'Record'),
                      style: context.type.titleHero.copyWith(color: colors.ink),
                    ),
                  ),
                  SizedBox(
                    width: 160,
                    child: SegmentedControl(
                      options: [
                        context.tr('সপ্তাহ', 'Week'),
                        context.tr('মাস', 'Month'),
                      ],
                      selectedIndex: _weekly ? 0 : 1,
                      onSelect: (i) => setState(() => _weekly = i == 0),
                    ),
                  ),
                ],
              ),
            ),
          ),
          Expanded(
            child: RepoBuilder<_RecordData>(
              key: ValueKey(_days),
              query: (repo) async => _RecordData(
                window: await repo.adherenceOver(_days),
                cells: await repo.dayCells(_days),
                perMed: await repo.perMedicineAdherence(_days),
              ),
              builder: (context, data) => SingleChildScrollView(
                padding: const EdgeInsets.fromLTRB(
                  Dimens.screenPadding,
                  0,
                  Dimens.screenPadding,
                  24,
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    _SummaryCard(window: data.window),
                    const SizedBox(height: 18),
                    _DayGrid(cells: data.cells),
                    const SizedBox(height: 18),
                    for (final m in data.perMed.where((m) => m.total > 0)) ...[
                      _MedBar(m: m),
                      const SizedBox(height: 18),
                    ],
                    SecondaryButton(
                      text: context.tr(
                        'ডাক্তারের জন্য পিডিএফ বানান',
                        'Make a PDF for the doctor',
                      ),
                      onPressed: context.nav.openDoctorReport,
                      height: 50,
                      enabled: data.window.total > 0,
                    ),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _RecordData {
  const _RecordData({required this.window, required this.cells, required this.perMed});

  final AdherenceWindow window;
  final List<DayCell> cells;
  final List<MedAdherence> perMed;
}

class _SummaryCard extends StatelessWidget {
  const _SummaryCard({required this.window});

  final AdherenceWindow window;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;

    // With nothing counted yet, a 0% ring reads as a failing grade rather than an empty record.
    if (window.total == 0) {
      return NbCard(
        padding: 13,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              context.tr('এখনও কিছু হিসাব হয়নি', 'Nothing counted yet'),
              style: context.type.cardTitlePrimary.copyWith(color: colors.ink),
            ),
            const SizedBox(height: 4),
            Text(
              context.tr(
                'ওষুধ খাওয়া শুরু করলে এখানে হিসাব জমতে থাকবে।',
                'Your record builds up here once you start confirming doses.',
              ),
              style: context.type.body.copyWith(color: colors.ink2),
            ),
          ],
        ),
      );
    }

    final pct = window.percent;
    return NbCard(
      padding: 13,
      child: Row(
        children: [
          ProgressRing(
            fraction: pct / 100,
            diameter: 74,
            strokeWidth: 8,
            center: Text(
              context.percent(pct),
              style: context.type.cardTitlePrimary.copyWith(
                fontSize: 20,
                color: colors.calmD,
              ),
            ),
          ),
          const SizedBox(width: 16),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  pct >= 90
                      ? context.tr('বেশ ভালো চলছে', 'Going well')
                      : pct >= 70
                          ? context.tr('মোটামুটি চলছে', 'Fairly steady')
                          : context.tr('আরও যত্ন দরকার', 'Needs more care'),
                  style: context.type.cardTitlePrimary.copyWith(color: colors.ink),
                ),
                Text(
                  window.missed == 0
                      ? context.tr(
                          'এই সময়ে কোনো ডোজ বাদ পড়েনি।',
                          'No doses were missed in this period.',
                        )
                      : context.tr(
                          'এই সময়ে ${context.num(window.missed)}টি ডোজ বাদ পড়েছে।',
                          '${window.missed} doses were missed in this period.',
                        ),
                  style: context.type.meta.copyWith(color: colors.ink2),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

const List<String> _bnWeekHeaders = ['সোম', 'মঙ্গল', 'বুধ', 'বৃহ', 'শুক্র', 'শনি', 'রবি'];
const List<String> _enWeekHeaders = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

class _DayGrid extends StatelessWidget {
  const _DayGrid({required this.cells});

  final List<DayCell> cells;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final headers = context.isBangla ? _bnWeekHeaders : _enWeekHeaders;
    final today = dateOnly(DateTime.now());

    // Chunk cells into weeks of 7. Pad the first week so weekday columns align (Mon = column 0).
    final firstDow = cells.isEmpty ? 0 : cells.first.date.weekday - 1;
    final padded = <DayCell?>[...List<DayCell?>.filled(firstDow, null), ...cells];
    final weeks = <List<DayCell?>>[];
    for (var i = 0; i < padded.length; i += 7) {
      weeks.add(padded.sublist(i, i + 7 > padded.length ? padded.length : i + 7));
    }

    return NbCard(
      padding: 13,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Row(
            children: [
              for (final h in headers)
                Expanded(
                  child: Center(
                    child: Text(
                      h,
                      style: context.type.meta.copyWith(fontSize: 12, color: colors.ink3),
                    ),
                  ),
                ),
            ],
          ),
          const SizedBox(height: 6),
          for (final week in weeks)
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 3),
              child: Row(
                children: [
                  for (var i = 0; i < 7; i++) ...[
                    if (i > 0) const SizedBox(width: 6),
                    Expanded(
                      child: AspectRatio(
                        aspectRatio: 1,
                        child: i < week.length && week[i] != null
                            ? _DayCellView(
                                cell: week[i]!,
                                isToday: week[i]!.date == today,
                              )
                            : const SizedBox.shrink(),
                      ),
                    ),
                  ],
                ],
              ),
            ),
          const SizedBox(height: 10),
          Wrap(
            spacing: 12,
            runSpacing: 6,
            children: [
              _LegendSwatch(color: colors.calm, label: context.tr('সব নেওয়া', 'All taken')),
              _LegendSwatch(color: colors.warm, label: context.tr('বাদ', 'Missed')),
              _LegendSwatch(color: colors.sage, label: context.tr('বাকি', 'Future')),
              _LegendSwatch(
                color: colors.calm,
                label: context.tr('কিছুটা', 'Some taken'),
                half: true,
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _DayCellView extends StatelessWidget {
  const _DayCellView({required this.cell, required this.isToday});

  final DayCell cell;
  final bool isToday;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final base = switch (cell.state) {
      DayState.full => colors.calm,
      DayState.missed => colors.warm,
      _ => colors.sage,
    };
    return DecoratedBox(
      decoration: BoxDecoration(
        color: base,
        borderRadius: BorderRadius.circular(9),
        // The "today" ring sits outside the cell, so it must not shrink the fill.
        border: isToday ? Border.all(color: colors.ink, width: 2.5) : null,
      ),
      child: cell.state == DayState.partial
          // 135° diagonal split: calm top-left triangle over the sage base.
          ? ClipRRect(
              borderRadius: BorderRadius.circular(9),
              child: CustomPaint(
                painter: _DiagonalHalfPainter(colors.calm),
                child: const SizedBox.expand(),
              ),
            )
          : const SizedBox.expand(),
    );
  }
}

class _DiagonalHalfPainter extends CustomPainter {
  const _DiagonalHalfPainter(this.color);

  final Color color;

  @override
  void paint(Canvas canvas, Size size) {
    final path = Path()
      ..moveTo(0, 0)
      ..lineTo(size.width, 0)
      ..lineTo(0, size.height)
      ..close();
    canvas.drawPath(path, Paint()..color = color);
  }

  @override
  bool shouldRepaint(_DiagonalHalfPainter old) => old.color != color;
}

class _LegendSwatch extends StatelessWidget {
  const _LegendSwatch({required this.color, required this.label, this.half = false});

  final Color color;
  final String label;
  final bool half;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        SizedBox(
          width: 12,
          height: 12,
          child: ClipRRect(
            borderRadius: BorderRadius.circular(4),
            child: ColoredBox(
              color: half ? colors.sage : color,
              // Mirrors the grid cell's 135° split so the legend matches what's drawn.
              child: half
                  ? CustomPaint(painter: _DiagonalHalfPainter(color))
                  : const SizedBox.expand(),
            ),
          ),
        ),
        const SizedBox(width: 5),
        Text(label, style: context.type.meta.copyWith(color: colors.ink3)),
      ],
    );
  }
}

class _MedBar extends StatelessWidget {
  const _MedBar({required this.m});

  final MedAdherence m;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final good = m.percent >= 80;
    return Row(
      children: [
        MedicineMark(shape: m.medicine.mark, color: Color(m.medicine.markColor), size: 20),
        const SizedBox(width: 10),
        SizedBox(
          width: 96,
          child: Text(
            m.medicine.displayName,
            style: context.type.meta.copyWith(color: colors.ink2),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
        ),
        const SizedBox(width: 8),
        Expanded(
          child: ClipRRect(
            borderRadius: BorderRadius.circular(5),
            child: SizedBox(
              height: 9,
              child: ColoredBox(
                color: colors.sage,
                child: FractionallySizedBox(
                  alignment: Alignment.centerLeft,
                  widthFactor: (m.percent / 100).clamp(0.0, 1.0),
                  child: ColoredBox(color: good ? colors.calm : colors.warm),
                ),
              ),
            ),
          ),
        ),
        const SizedBox(width: 8),
        Text(context.percent(m.percent), style: context.type.meta.copyWith(color: colors.ink2)),
      ],
    );
  }
}
